File: app/src/main/java/ezike/tobenna/myweather/data/local/WeatherTypeConverter.java
package ezike.tobenna.myweather.data.local;

import com.squareup.moshi.Moshi;

import java.io.IOException;

import androidx.room.TypeConverter;
import ezike.tobenna.myweather.data.model.CurrentWeather;
import ezike.tobenna.myweather.data.model.Forecast;
import ezike.tobenna.myweather.data.model.WeatherLocation;

/**
 * @author tobennaezike
 */
public class WeatherTypeConverter {

    private Moshi moshi = new Moshi.Builder().build();

    @TypeConverter
    public String weatherToJson(CurrentWeather weather) {
        return weather == null ? null : moshi.adapter(CurrentWeather.class).toJson(weather);
    }

    @TypeConverter
    public CurrentWeather weatherFromJson(String string) {
        CurrentWeather weather;
        try {
            weather = moshi.adapter(CurrentWeather.class).fromJson(string);
        } catch (IOException e) {
            e.printStackTrace();
            return null;
        }
        return weather;
    }


    @TypeConverter
    public String locationToJson(WeatherLocation weatherLocation) {
        return weatherLocation == null ? null : moshi.adapter(WeatherLocation.class).toJson(weatherLocation);
    }

    @TypeConverter
    public WeatherLocation locationFromJson(String string) {
        WeatherLocation location;
        try {
            location = moshi.adapter(WeatherLocation.class).fromJson(string);
        } catch (IOException e) {
            e.printStackTrace();
            return null;
        }
        return location;
    }


    @TypeConverter
    public String forecastToJson(Forecast forecast) {
        return forecast == null ? null : moshi.adapter(Forecast.class).toJson(forecast);
    }

    @TypeConverter
    public Forecast forecastFromJson(String string) {
        Forecast weatherForecast;
        try {
            weatherForecast = moshi.adapter(Forecast.class).fromJson(string);
        } catch (IOException e) {
            e.printStackTrace();
            return null;
        }
        return weatherForecast;
    }
}

File: app/src/main/java/ezike/tobenna/myweather/data/local/db/WeatherTypeConverter.kt
package ezike.tobenna.myweather.data.local.db

import androidx.room.TypeConverter
import com.squareup.moshi.Moshi.Builder
import ezike.tobenna.myweather.data.model.Current
import ezike.tobenna.myweather.data.model.Request
import ezike.tobenna.myweather.data.model.WeatherLocation
import ezike.tobenna.myweather.data.model.WeatherResponse
import java.io.IOException

/**
 * @author tobennaezike
 */
class WeatherTypeConverter {
    private val moshi = Builder().build()

    @TypeConverter
    fun weatherToJson(weather: WeatherResponse?): String? {
        return if (weather == null) null else moshi.adapter<Any>(WeatherResponse::class.java).toJson(weather)
    }

    @TypeConverter
    fun weatherFromJson(string: String): WeatherResponse? {
        val weather: WeatherResponse?
        weather = try {
            moshi.adapter<WeatherResponse>(WeatherResponse::class.java).fromJson(string)
        } catch (e: IOException) {
            e.printStackTrace()
            return null
        }
        return weather
    }

    @TypeConverter
    fun locationToJson(weatherLocation: WeatherLocation?): String? {
        return if (weatherLocation == null) null else moshi.adapter<Any>(WeatherLocation::class.java).toJson(weatherLocation)
    }

    @TypeConverter
    fun locationFromJson(string: String?): WeatherLocation? {
        val location: WeatherLocation?
        location = try {
            moshi.adapter<WeatherLocation>(WeatherLocation::class.java).fromJson(string)
        } catch (e: IOException) {
            e.printStackTrace()
            return null
        }
        return location
    }

    @TypeConverter
    fun currentToJson(current: Current?): String? {
        return if (current == null) null else moshi.adapter<Any>(Current::class.java).toJson(current)
    }

    @TypeConverter
    fun currentFromJson(string: String): Current? {
        val current: Current?
        current = try {
            moshi.adapter<Current>(Current::class.java).fromJson(string)
        } catch (e: IOException) {
            e.printStackTrace()
            return null
        }
        return current
    }

    @TypeConverter
    fun requestToJson(request: Request?): String? {
        return if (request == null) null else moshi.adapter<Any>(Request::class.java).toJson(request)
    }

    @TypeConverter
    fun requestFromJson(string: String): Request? {
        val request: Request?
        request = try {
            moshi.adapter<Request>(Request::class.java).fromJson(string)
        } catch (e: IOException) {
            e.printStackTrace()
            return null
        }
        return request
    }
}